### ISSUE TITLE
clang-tidy crashes with __put_user_goto

### ISSUE THREAD

**Issue by tmroeder** [ClangBuiltLinux/linux#519]
I was trying to apply clang-tidy on the x86 kernel using a recent clang and a recent kernel, and clang-tidy crashes. I bisected the kernel to the offending commit, and it looks like it has something to do with asm-goto. I don't know much about the asm-goto work, so I'm posting the bug here.

The relevant versions are

$ ~/.local/toolchains/clang_head/release/bin/clang --version
clang version 9.0.0 (https://github.com/llvm/llvm-project.git 53cd7406bb6507f60c14fbbce4b213f0bbd78730)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /usr/local/google/home/tmroeder/.local/toolchains/clang_head/release/bin

git bisect gave me the failing commit as

$ git bisect good
4a789213c9a54c8b618924d3421e56e98df8a447 is the first bad commit
commit 4a789213c9a54c8b618924d3421e56e98df8a447
Author: Linus Torvalds <torvalds@linux-foundation.org>
Date:   Sat May 21 21:44:51 2016 -0700

    x86 uaccess: Introduce __put_user_goto

    This is finally the actual reason for the odd error handling in the
    "unsafe_get/put_user()" functions, introduced over three years ago.

    Using a "jump to error label" interface is somewhat odd, but very
    convenient as a programming interface, and more importantly, it fits
    very well with simply making the target be the exception handler address
    directly from the inline asm.

    The reason it took over three years to actually do this? We need "asm
    goto" support for it, which only became the default on x86 last year.
    It's now been a year that we've forced asm goto support (see commit
    e501ce957a78 "x86: Force asm-goto"), and so let's just do it here too.

    [ Side note: this commit was originally done back in 2016. The above
      commentary about timing is obviously about it only now getting merged
      into my real upstream tree     - Linus ]

    Sadly, gcc still only supports "asm goto" with asms that do not have any
    outputs, so we are limited to only the put_user case for this.  Maybe in
    several more years we can do the get_user case too.

    Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>

 arch/x86/include/asm/uaccess.h | 28 +++++++++++++++++-----------
 1 file changed, 17 insertions(+), 11 deletions(-)

This commit was originally created in 2016, but it wasn't added to the upstream tree until Jan 2019.

Here's a backtrace from the crash

```
$ ~/.local/toolchains/clang_head/debug/bin/clang-tidy -p . fs/eventpoll.c
Analysis for this terminator not implemented.
UNREACHABLE executed at /usr/local/google/home/tmroeder/src/llvm/llvm-project/clang/lib/StaticAnalyzer/Core/CoreEngine.cpp:300!
 #0 0x0000562e294fb77a llvm::sys::PrintStackTrace(llvm::raw_ostream&) /usr/local/google/home/tmroeder/src/llvm/llvm-project/llvm/lib/Support/Unix/Signals.inc:494:0
 #1 0x0000562e294fb80d PrintStackTraceSignalHandler(void*) /usr/local/google/home/tmroeder/src/llvm/llvm-project/llvm/lib/Support/Unix/Signals.inc:558:0
 #2 0x0000562e294f9acb llvm::sys::RunSignalHandlers() /usr/local/google/home/tmroeder/src/llvm/llvm-project/llvm/lib/Support/Signals.cpp:68:0
 #3 0x0000562e294fb26d SignalHandler(int) /usr/local/google/home/tmroeder/src/llvm/llvm-project/llvm/lib/Support/Unix/Signals.inc:357:0
 #4 0x00007fcd2ed633a0 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x123a0)
 #5 0x00007fcd2c43ecfb raise (/lib/x86_64-linux-gnu/libc.so.6+0x36cfb)
 #6 0x00007fcd2c4298ad abort (/lib/x86_64-linux-gnu/libc.so.6+0x218ad)
 #7 0x0000562e29468dff bindingsErrorHandler(void*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool) /usr/local/google/home/tmroeder/src/llvm/llvm-project/llvm/lib/Support/ErrorHandling.cpp:230:0
 #8 0x0000562e2a6d5950 clang::ento::CoreEngine::HandleBlockExit(clang::CFGBlock const*, clang::ento::ExplodedNode*) /usr/local/google/home/tmroeder/src/llvm/llvm-project/clang/lib/StaticAnalyzer/Core/CoreEngine.cpp:303:0
 #9 0x0000562e2a6d6409 clang::ento::CoreEngine::HandlePostStmt(clang::CFGBlock const*, unsigned int, clang::ento::ExplodedNode*) /usr/local/google/home/tmroeder/src/llvm/llvm-project/clang/lib/StaticAnalyzer/Core/CoreEngine.cpp:464:0
#10 0x0000562e2a6d504b clang::ento::CoreEngine::dispatchWorkItem(clang::ento::ExplodedNode*, clang::ProgramPoint, clang::ento::WorkListUnit const&) /usr/local/google/home/tmroeder/src/llvm/llvm-project/clang/lib/StaticAnalyzer/Core/CoreEngine.cpp:193:0
#11 0x0000562e2a6d4cbe clang::ento::CoreEngine::ExecuteWorkList(clang::LocationContext const*, unsigned int, llvm::IntrusiveRefCntPtr<clang::ento::ProgramState const>) /usr/local/google/home/tmroeder/src/llvm/llvm-project/clang/lib/StaticAnalyzer/Core/CoreEngine.cpp:148:0
#12 0x0000562e2a19e623 clang::ento::ExprEngine::ExecuteWorkList(clang::LocationContext const*, unsigned int) /usr/local/google/home/tmroeder/src/llvm/llvm-project/clang/include/clang/StaticAnalyzer/Core/PathSensitive/ExprEngine.h:168:0
#13 0x0000562e2a14ddd8 (anonymous namespace)::AnalysisConsumer::RunPathSensitiveChecks(clang::Decl*, clang::ento::ExprEngine::InliningModes, llvm::DenseSet<clang::Decl const*, llvm::DenseMapInfo<clang::Decl const*> >*) /usr/local/google/home/tmroeder/src/llvm/llvm-project/clang/lib/StaticAnalyzer/Frontend/AnalysisConsumer.cpp:772:0
#14 0x0000562e2a14dc36 (anonymous namespace)::AnalysisConsumer::HandleCode(clang::Decl*, unsigned int, clang::ento::ExprEngine::InliningModes, llvm::DenseSet<clang::Decl const*, llvm::DenseMapInfo<clang::Decl const*> >*) /usr/local/google/home/tmroeder/src/llvm/llvm-project/clang/lib/StaticAnalyzer/Frontend/AnalysisConsumer.cpp:745:0
#15 0x0000562e2a14cb19 (anonymous namespace)::AnalysisConsumer::HandleDeclsCallGraph(unsigned int) /usr/local/google/home/tmroeder/src/llvm/llvm-project/clang/lib/StaticAnalyzer/Frontend/AnalysisConsumer.cpp:539:0
#16 0x0000562e2a14ceb1 (anonymous namespace)::AnalysisConsumer::runAnalysisOnTranslationUnit(clang::ASTContext&) /usr/local/google/home/tmroeder/src/llvm/llvm-project/clang/lib/StaticAnalyzer/Frontend/AnalysisConsumer.cpp:585:0
#17 0x0000562e2a14d0af (anonymous namespace)::AnalysisConsumer::HandleTranslationUnit(clang::ASTContext&) /usr/local/google/home/tmroeder/src/llvm/llvm-project/clang/lib/StaticAnalyzer/Frontend/AnalysisConsumer.cpp:616:0
#18 0x0000562e2a91cdcd clang::MultiplexConsumer::HandleTranslationUnit(clang::ASTContext&) /usr/local/google/home/tmroeder/src/llvm/llvm-project/clang/lib/Frontend/MultiplexConsumer.cpp:291:0
#19 0x0000562e2ab21ebd clang::ParseAST(clang::Sema&, bool, bool) /usr/local/google/home/tmroeder/src/llvm/llvm-project/clang/lib/Parse/ParseAST.cpp:178:0
#20 0x0000562e2a8ea0d3 clang::ASTFrontendAction::ExecuteAction() /usr/local/google/home/tmroeder/src/llvm/llvm-project/clang/lib/Frontend/FrontendAction.cpp:1035:0
#21 0x0000562e2a8e9acb clang::FrontendAction::Execute() /usr/local/google/home/tmroeder/src/llvm/llvm-project/clang/lib/Frontend/FrontendAction.cpp:938:0
#22 0x0000562e2a88f204 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) /usr/local/google/home/tmroeder/src/llvm/llvm-project/clang/lib/Frontend/CompilerInstance.cpp:945:0
#23 0x0000562e2a0deaa5 clang::tooling::FrontendActionFactory::runInvocation(std::shared_ptr<clang::CompilerInvocation>, clang::FileManager*, std::shared_ptr<clang::PCHContainerOperations>, clang::DiagnosticConsumer*) /usr/local/google/home/tmroeder/src/llvm/llvm-project/clang/lib/Tooling/Tooling.cpp:369:0
#24 0x0000562e29a0b315 clang::tidy::runClangTidy(clang::tidy::ClangTidyContext&, clang::tooling::CompilationDatabase const&, llvm::ArrayRef<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, llvm::IntrusiveRefCntPtr<llvm::vfs::OverlayFileSystem>, bool, llvm::StringRef)::ActionFactory::runInvocation(std::shared_ptr<clang::CompilerInvocation>, clang::FileManager*, std::shared_ptr<clang::PCHContainerOperations>, clang::DiagnosticConsumer*) /usr/local/google/home/tmroeder/src/llvm/llvm-project/clang-tools-extra/clang-tidy/ClangTidy.cpp:541:0
#25 0x0000562e2a0de8fe clang::tooling::ToolInvocation::runInvocation(char const*, clang::driver::Compilation*, std::shared_ptr<clang::CompilerInvocation>, std::shared_ptr<clang::PCHContainerOperations>) /usr/local/google/home/tmroeder/src/llvm/llvm-project/clang/lib/Tooling/Tooling.cpp:345:0
#26 0x0000562e2a0de72d clang::tooling::ToolInvocation::run() /usr/local/google/home/tmroeder/src/llvm/llvm-project/clang/lib/Tooling/Tooling.cpp:330:0
#27 0x0000562e2a0dfd26 clang::tooling::ClangTool::run(clang::tooling::ToolAction*) /usr/local/google/home/tmroeder/src/llvm/llvm-project/clang/lib/Tooling/Tooling.cpp:518:0
#28 0x0000562e29a0b65f clang::tidy::runClangTidy(clang::tidy::ClangTidyContext&, clang::tooling::CompilationDatabase const&, llvm::ArrayRef<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, llvm::IntrusiveRefCntPtr<llvm::vfs::OverlayFileSystem>, bool, llvm::StringRef) /usr/local/google/home/tmroeder/src/llvm/llvm-project/clang-tools-extra/clang-tidy/ClangTidy.cpp:562:0
#29 0x0000562e28f14342 clang::tidy::clangTidyMain(int, char const**) /usr/local/google/home/tmroeder/src/llvm/llvm-project/clang-tools-extra/clang-tidy/tool/ClangTidyMain.cpp:431:0
#30 0x0000562e28f1482b main /usr/local/google/home/tmroeder/src/llvm/llvm-project/clang-tools-extra/clang-tidy/tool/ClangTidyMain.cpp:492:0
#31 0x00007fcd2c42b52b __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2352b)
#32 0x0000562e28f12bfa _start (/usr/local/google/home/tmroeder/.local/toolchains/clang_head/debug/bin/clang-tidy+0x746bfa)
Aborted
```
repro steps from the bad commit:

```
$ make defconfig
$ make CC=/path/to/clang
$ scripts/gen_compile_commands.py
$ /path/to/clang-tidy -p . fs/eventpoll.c
```

**Comment by tmroeder** [comment]
Looking at the kernel commit and the assertion, I assume that this is because asm goto introduces a new kind of terminator that the static analyzer doesn't know about.

**Comment by nickdesaulniers** [comment]
`clang::ento::CoreEngine::HandleBlockExit` (or hidden inlined functions) probably don't understand the newly added `callbr` LLVM IR instruction. cc @topperc (I doubt this is an issue in the kernel, so the bisection results are a red herring).

**Comment by tmroeder** [comment]
I'm confused: what does Static Analyzer code (clang/lib/StaticAnalyzer/Core/CoreEngine.cpp:303) have to do with IR? I thought the Static Analyzer operated above the level of IR entirely.

**Comment by nickdesaulniers** [comment]
> Analysis for this terminator not implemented.
> UNREACHABLE executed at /usr/local/google/home/tmroeder/src/llvm/llvm-project/clang/lib/StaticAnalyzer/Core/CoreEngine.cpp:300!

Looks like it operates on ASTNodes, so `Stmt::GCCAsmStmt`  where `isAsmGoto() == true` needs support.

**Comment by tmroeder** [comment]
I can confirm that the terminator is a GCCAsmStmt: I added a dump statement for the terminator in the function, right before the UNREACHABLE error, and I got:

```
GCCAsmStmt 0x559a3a37e1b0
|-ParenExpr 0x559a3a37dfc8 'typeof (*(&uevent->events))':'unsigned int'
| `-CStyleCastExpr 0x559a3a37dfa0 'typeof (*(&uevent->events))':'unsigned int' <NoOp>
|   `-ImplicitCastExpr 0x559a3a37df88 '__poll_t':'unsigned int' <LValueToRValue> part_of_explicit_cast
|     `-ParenExpr 0x559a3a37df68 '__poll_t':'unsigned int' lvalue
|       `-DeclRefExpr 0x559a3a37def0 '__poll_t':'unsigned int' lvalue Var 0x559a3a37aef0 'revents' '__poll_t':'unsigned int'
`-ParenExpr 0x559a3a37e170 'struct __large_struct':'struct __large_struct' lvalue
  `-UnaryOperator 0x559a3a37e158 'struct __large_struct':'struct __large_struct' lvalue prefix '*' cannot overflow
    `-CStyleCastExpr 0x559a3a37e130 'struct __large_struct *' <BitCast>
      `-ParenExpr 0x559a3a37e110 '__poll_t *'
        `-ParenExpr 0x559a3a37e0a8 '__poll_t *'
          `-ParenExpr 0x559a3a37e088 '__poll_t *'
            `-UnaryOperator 0x559a3a37e070 '__poll_t *' prefix '&' cannot overflow
              `-MemberExpr 0x559a3a37e040 '__poll_t':'unsigned int' lvalue ->events 0x559a39470f50
                `-ImplicitCastExpr 0x559a3a37e028 'struct epoll_event *' <LValueToRValue>
                  `-DeclRefExpr 0x559a3a37e008 'struct epoll_event *' lvalue Var 0x559a3a37b040 'uevent' 'struct epoll_event *'
```

**Comment by nickdesaulniers** [comment]
right, one issue is that `asm` and `asm goto` are both `Stmt::GCCAsmStmt`, but only `asm goto` is a terminator (which implies control flow), which can be checked by calling `Stmt::isAsmGoto()`.

**Comment by tmroeder** [comment]
Looking back, I see why the kernel bisection might have been confusing. I didn't think there was anything wrong with the kernel. Rather, I used to be able to run clang-tidy on arbitrary kernel code before asm-goto got added, and I was trying to find the point in the kernel when clang-tidy stopped working. The bisection (using a HEAD version of clang) showed that it was when a particular use of asm-goto got turned on.

In retrospect, I should have just added the print statement to see what the terminator was that was causing the Static Analyzer to crash.

FYI, for completeness, I added the isAsmGoto checked that you suggested, and it is indeed an asm goto (as expected from all the rest of the information in this bug). I don't know that part of the code well, but I can take a look to see if I can figure out what to do.
